share/man/man4/coreindex.4
.\"
.\" Copyright (c) 2026 Koine Yuusuke
.\"
.\" SPDX-License-Identifier: BSD-2-Clause
.\"
.Dd Jul 2, 2026
.Dt COREINDEX 4
.Os
.Sh NAME
.Nm coreindex
.Nd CPU core performance/efficiency/capacity manual setting driver
.Sh SYNOPSIS
To compile this driver into the kernel,
place the following line in your
kernel configuration file:
.Bd -ragged -offset indent
.Cd "device coreindex"
.Ed
.Pp
Alternatively, to load the driver as a
module at boot time, place the following line in
.Xr loader.conf 5 :
.Bd -literal -offset indent
coreindex_load="YES"
.Ed
.Pp
Do not both compile the driver into the kernel and
load it as a module at boot time.
.Pp
This driver can also cooperate with
.Xr hmp 4
to pass internal performance metrics to the ULE scheduler.
To enable this capability, add the following line to
your kernel configuration file:
.Bd -ragged -offset indent
.Cd "options HMP"
.Ed
.Sh DESCRIPTION
The
.Nm
driver allows for the manual configuration of 'capacity',
'performance', and 'efficiency' scores for each CPU core
managed by the heterogeneous multi-core module hmp(4).
By using this driver to set appropriate values based on
actual CPU core performance, it becomes possible to have
.Xr sched_ule 4
perform scheduling that accounts for per-core performance
characteristics in scenarios such as the following:
.Bl -bullet
.It
Heterogeneous multi-core environments where hardware-level
retrieval of per-core performance values is not supported
.It
Environments where no hmp(4) provider driver is available.
.It
Heterogeneous multi-core environments that support
hardware-level performance reporting, but where manual
configuration of per-core performance values is preferred
.It
Debugging within virtualized environments
.El

Per-core 'capacity', 'performance', and 'efficiency' values
are configured via the sysctl interface and
.Xr loader.conf 5
, as described below.

Note that this driver does not support dynamic loading via
the
.Xr kldload 8
command.

.Sh SYSCTL VARIABLES AND LOADER TUNABLES

The behavior of
.Nm
is influenced by the following
.Xr sysctl 8
and
.Xr loader.conf 5
tunables:
.Bl -tag -width indent
.It Va hw.coreindex.%d.capacity
Set a value representing the relative throughput of the target
CPU core. This value should account for both IPC (Instructions
Per Cycle) and maximum clock frequency, rather than simply
reflecting the clock speed ratio.

The value range is 1 to 1024, with higher values indicating
higher performance. No normalization is required if the value
falls within this 1 to 1024 range.

If no 'capacity' value is set, the 'capacity' value for all CPU
cores will be set to 1024. If even one 'capacity' value is set,
the 'capacity' values for the remaining unset CPU cores will be
set to the midpoint between the maximum and minimum values
manually set.

This value can only be set at boot time via
.Xr loader.conf 5
\&. After the system has booted, the value can only be viewed via
.Xr sysctl 5
; it cannot be modified.
.It Va hw.coreindex.%d.performance
This is a dynamic value representing the current processing
capability of a given CPU core. A higher value is assigned to
CPU cores capable of delivering higher performance at a given
moment, taking thermal and power constraints into account.
This value reflects current conditions rather than intrinsic
capabilities such as IPC or maximum clock frequency. For
instance, when using a driver like
.Xr intelhfi 4
, the value is dynamically updated based on data provided by
the CPU itself.

However, when using
.Xr coreindex 4
, it is not possible to monitor real-time thermal or power
conditions and manually update the value accordingly.
Therefore, the standard approach is to assign higher values
to the CPU cores you wish to prioritize for active use.

The value must be set within the range of 0 to 1024; no
normalization is required.

Values can be configured in advance via
.Xr loader.conf 5
or set after system startup using
.Xr sysctl 8
\&. Dynamic updates via
.Xr sysctl 8
are also supported.

CPU cores for which no value is specified default to the
maximum value of 1024.
.It Va hw.coreindex.%d.efficiency
This is a dynamic value representing the current power-saving
metric of the target CPU core. A higher value is assigned to
a CPU core that can operate in a power-saving mode at that
specific moment, based on thermal and power considerations.
This value reflects the conditions at that particular time
and does not represent the target CPU core's base clock
frequency or similar specifications. For instance, when using
a driver like
.Xr intelhfi 4
, the value is dynamically updated based on data provided by
the CPU itself.

However, when using
.Xr coreindex 4
, it is not possible to monitor CPU thermal or power
conditions and manually update the settings in real-time
based on that information.
Therefore, the standard approach is to assign higher values
to the CPU cores you wish to prioritize for active use, while
operating under the general premise of conserving power.

The value must be set within the range of 0 to 1024; no
normalization is required.

Values can be configured in advance via
.Xr loader.conf 5
or set after system startup using
.Xr sysctl 8
\&. Dynamic updates via
.Xr sysctl 8
are also supported.

CPU cores for which no value is specified default to the
maximum value of 1024.
.Sh SEE ALSO
.Xr hmp 4 ,
.Xr sched_ule 4 ,
.Xr intelhfi 4 ,
.Xr loader.conf 5 ,
.Xr sysctl 8
.Sh HISTORY
The
.Nm
driver first appeared in FreeBSD 16.0.
.Sh AUTHORS
.An -nosplit
The
.Nm
driver was written by
.An Koine Yuusuke Aq Mt koinec@yahoo.co.jp .


sys/dev/coreindex/coreindex.c
/*
 * Copyright (c) 2026 Koine Yuusuke
 *
 * SPDX-License-Identifier: BSD-2-Clause
 */

#include <sys/cdefs.h>
#include "opt_global.h"
#include "opt_coreindex.h"

#include <sys/malloc.h>
#include <sys/bus.h>
#include <sys/kernel.h>
#include <sys/module.h>
#include <sys/mutex.h>
#include <sys/lock.h>
#include <sys/proc.h>
#include <sys/smp.h>
#include <sys/hmp.h>
#include <sys/sysctl.h>

#define	COREINDEX_NAME "coreindex"
#define	COREINDEX_HMP_PRIORITY 10

/* Check kernel build option with SMP */
#ifndef SMP
#error "coreindex requires the SMP option in the kernel configuraton."
#endif

/* Check kernel build option with HMP */
#ifndef HMP
#error "coreindex requires the HMP option in the kernel configuraton."
#endif

/*
 * Structure defines
 */
struct coreindex_instance {
	uint8_t		status;

	uint32_t	*capacity;

	struct {
		uint16_t	*performance;
		uint16_t	*efficiency;
	} score;

	struct {
		struct sysctl_ctx_list ctx;
		struct sysctl_oid *parent;
		struct sysctl_oid **core;
	} sysctl;
};

#define	COREINDEX_STATUS_LOADED	(1 << 0)	/* status bit0: coreindex loaded */

static struct coreindex_instance cidx_inst;

/*
 * Prototype defines
 */
static uint32_t coreindex_capacity_get_tunable_value(void);
static int coreindex_capacity_provider_probe(void);
static int coreindex_capacity_provider_init(void);
static uint16_t coreindex_score_get_tunable(char *nodename, uint16_t *valptr);
static int coreindex_score_provider_probe(void);
static int coreindex_score_provider_init(void);
static int coreindex_modevent(module_t mod, int type, void *data);
static int coreindex_register_sysctl(void);
static int coreindex_capacity_register_sysctl(void);
static int coreindex_score_set_sysctl(SYSCTL_HANDLER_ARGS);
static int coreindex_score_register_sysctl(void);

DPCPU_DECLARE(struct hmp_pcpu, hmp_pcpu);

static MALLOC_DEFINE(M_COREINDEX, COREINDEX_NAME, "Buffers for coreindex driver");

/*
 * HMP capacity provider
 */
static uint32_t
coreindex_capacity_get_tunable_value(void)
{
	char path[128];
	uint64_t value;
	uint32_t mincap = HMP_CAPACITY_SCALE;
	uint32_t maxcap = 0;

	/* Set the capacity value of the core for which tunable parameters are configured. */
	for (int i=0; i<mp_ncpus; i++) {
		/* Set temporary (not initialized) value */
		*(cidx_inst.capacity + i) = 0;

		/* Get tunable capacity value */
		snprintf(path, sizeof(path), "hw.coreindex.%d.capacity", i);
		if (TUNABLE_UINT64_FETCH(path, &value)) {
			if ((value > 0) && (value <= HMP_CAPACITY_SCALE))
				*(cidx_inst.capacity + i) = (uint32_t)value;

			if (mincap > *(cidx_inst.capacity + i))
				mincap = *(cidx_inst.capacity + i);

			if (maxcap < *(cidx_inst.capacity + i))
				maxcap = *(cidx_inst.capacity + i);
		}
	}

	/* Calc. default capacity value */
	if ((mincap == HMP_CAPACITY_SCALE) && (maxcap == 0))
		value = HMP_CAPACITY_SCALE;
	else
		value = (mincap + maxcap) / 2;

	/* Set the capacity value of the core for which tunable parameters are NOT configured. */
	for (int i=0; i<mp_ncpus; i++) {
		if (*(cidx_inst.capacity + i) > 0)
			continue;

		/* Get tunable capacity value */
		*(cidx_inst.capacity + i) = value;
	}

	return mincap;
}

static int
coreindex_capacity_provider_probe(void)
{
	return (0);
}

static int
coreindex_capacity_provider_init(void)
{
	struct hmp_pcpu *hp;
	uint32_t mincap;
	uint32_t *capacity;
	int cpu;

	/* Alloc CPU core perf. value */
	capacity = malloc((sizeof(uint32_t) * mp_ncpus), M_COREINDEX, M_NOWAIT | M_ZERO);
	if (capacity == NULL)
		return (-1);

	cidx_inst.capacity = capacity;

	/* Get tunable capacity value */
	mincap = coreindex_capacity_get_tunable_value();

	/* Regist sysctl interfaces */
	if (coreindex_register_sysctl()) {
		free(cidx_inst.capacity, M_COREINDEX);
		return (-2);
	}

	/* Regist sysctl capacity leaf */
	coreindex_capacity_register_sysctl();

	/* Set CPU core capacity for HMP */
	CPU_FOREACH(cpu) {
		hp = DPCPU_ID_PTR(cpu, hmp_pcpu);
		hp->capacity = HMP_CAPACITY_NORMAL_FROM(*(cidx_inst.capacity + cpu), mincap);
	}

	/* Set coreindex loaded flag */
	cidx_inst.status |= COREINDEX_STATUS_LOADED;

	return (0);
}

static struct hmp_capacity_provider coreindex_capacity_provider = {
	.name		= COREINDEX_NAME,
	.priority	= COREINDEX_HMP_PRIORITY,
	.probe		= coreindex_capacity_provider_probe,
	.init		= coreindex_capacity_provider_init,
};
HMP_CAPACITY_PROVIDER_DECLARE(coreindex, coreindex_capacity_provider);

/*
 * HMP socre provider
 */
static uint16_t
coreindex_score_get_tunable(char *nodename, uint16_t *valptr)
{
	char path[128];
	uint64_t value;
	uint16_t maxval = 0;

	for (int i=0; i<mp_ncpus; i++) {
		/* Get tunable score value */
		snprintf(path, sizeof(path), "hw.coreindex.%d.%s", i, nodename);
		if (TUNABLE_UINT64_FETCH(path, &value))
			*(valptr + i) = (uint16_t)((value > HMP_SCORE_MAX) ?  HMP_SCORE_DEFAULT : value);
		else
			*(valptr + i) = HMP_SCORE_DEFAULT;

		if (maxval < *(valptr + i))
			maxval = *(valptr + i);
	}

	return maxval;
}

static int
coreindex_score_provider_probe(void)
{
	return (0);
}

static int
coreindex_score_provider_init(void)
{
	struct hmp_pcpu *hp;
	uint16_t *performance;
	uint16_t *efficiency;
	int cpu;

	/* Check capacity provider regist status */
	if (!(cidx_inst.status & COREINDEX_STATUS_LOADED))
		return (-1);

	/* Alloc CPU core perf. value */
	performance = malloc((sizeof(uint16_t) * mp_ncpus), M_COREINDEX, M_NOWAIT | M_ZERO);
	if (performance == NULL)
		return (-2);

	cidx_inst.score.performance = performance;

	/* Alloc CPU core efficiency value */
	efficiency = malloc((sizeof(uint16_t) * mp_ncpus), M_COREINDEX, M_NOWAIT | M_ZERO);
	if (efficiency == NULL) {
		free(cidx_inst.score.performance, M_COREINDEX);
		return (-3);
	}
	cidx_inst.score.efficiency = efficiency;

	/* Get tunable capacity value */
	coreindex_score_get_tunable("performance", performance);
	coreindex_score_get_tunable("efficiency", efficiency);

	/* Regist sysctl score leaf */
	coreindex_score_register_sysctl();

	/* Set CPU core capacity for HMP */
	CPU_FOREACH(cpu) {
		hp = DPCPU_ID_PTR(cpu, hmp_pcpu);
		hmp_set_score(hp, HMP_SCORE_PERF, *(cidx_inst.score.performance + cpu));
		hmp_set_score(hp, HMP_SCORE_EFF, *(cidx_inst.score.efficiency + cpu));
	}

	printf("coreindex: CPU core performance/efficiency index setter\n");

	return (0);
}

static struct hmp_score_provider coreindex_score_provider = {
	.name		= COREINDEX_NAME,
	.priority	= COREINDEX_HMP_PRIORITY,
	.probe		= coreindex_score_provider_probe,
	.init		= coreindex_score_provider_init,
};
HMP_SCORE_PROVIDER_DECLARE(coreindex, coreindex_score_provider);

/*
 * Module Interfaces
 */
static int
coreindex_modevent(module_t mod, int type, void *data)
{
	switch (type) {
	case MOD_LOAD:
		/*
		 * If this driver is enabled in the /boot/loader.conf file, it will be
		 * initialized as a driver module after initialization by hmp(4).
		 * If an error is returned in this case, an error message will be recorded
		 * in dmesg due to double loading, even though it has been loaded successfully.
		 * Therefore, if it has been initialized by hmp(4), zero is returned, and it will
		 * be treated as having been loaded without errors.
		 */
		if (cidx_inst.status & COREINDEX_STATUS_LOADED)
			return (0);

		/*
		 * To allow enabling the driver from the /boot/loader.conf file, it is registered
		 * as a driver module, but loading via the kldload command is not permitted.
		 * Therefore, it returns an error.
		 */
		return (EOPNOTSUPP);

	default:
		return (EOPNOTSUPP);
	}
}

static moduledata_t coreindex_mod = {
	COREINDEX_NAME,		/* module name */
	coreindex_modevent,	/* event handler */
	NULL
};
DECLARE_MODULE(coreindex, coreindex_mod, SI_SUB_SMP + 2, SI_ORDER_ANY);

/*
 * Sysctls
 */
static int
coreindex_register_sysctl(void)
{
	char id[8];
	char desc[128];
	struct sysctl_oid *node;

	sysctl_ctx_init(&(cidx_inst.sysctl.ctx));

	/* Regist "hw.coreindex" parent node */
	cidx_inst.sysctl.parent = SYSCTL_ADD_NODE( &(cidx_inst.sysctl.ctx),
			SYSCTL_STATIC_CHILDREN(_hw), OID_AUTO, COREINDEX_NAME,
			CTLFLAG_RD | CTLFLAG_MPSAFE, 0,
			"coreindex: CPU core performance/efficiency index setter");
	if (cidx_inst.sysctl.parent == NULL)
		return (-1);

	/* Alloc sysctl node memory */
	cidx_inst.sysctl.core = malloc((sizeof(struct sysctl_oid *) * mp_ncpus),
								M_COREINDEX, M_NOWAIT | M_ZERO);
	if (cidx_inst.sysctl.core == NULL) {
		sysctl_ctx_free(&(cidx_inst.sysctl.ctx));
		return (-2);
	}

	/* Regist "hw.coreindex.X" node each CPU core id */
	for (int i=0; i<mp_ncpus; i++) {
		snprintf(id, sizeof(id), "%d", i);
		snprintf(desc, sizeof(desc), "CPU core #%d performance/efficiency index", i);

		node = SYSCTL_ADD_NODE(&(cidx_inst.sysctl.ctx), SYSCTL_CHILDREN(cidx_inst.sysctl.parent),
					OID_AUTO, id, CTLFLAG_RD | CTLFLAG_MPSAFE, 0, desc);
		if (node == NULL) {
			sysctl_ctx_free(&(cidx_inst.sysctl.ctx));
			free(cidx_inst.sysctl.core, M_COREINDEX);
			return (-3);
		}

		*(cidx_inst.sysctl.core + i) = node;
	}

	return (0);
}

static int
coreindex_capacity_register_sysctl(void)
{
	char desc[128];

	/* Set sysctl capacity value */
	for (int i=0; i<mp_ncpus; i++) {
		snprintf(desc, sizeof(desc), "CPU core #%d capacity index", i);
		SYSCTL_ADD_U32( &(cidx_inst.sysctl.ctx), SYSCTL_CHILDREN(*(cidx_inst.sysctl.core+i)),
				OID_AUTO, "capacity", CTLFLAG_RDTUN, (cidx_inst.capacity + i), 0, desc);
	}

	return (0);
}

static int
coreindex_score_set_sysctl(SYSCTL_HANDLER_ARGS)
{
	struct hmp_pcpu *hp;
	uint16_t *valptr;
	int cpu = arg2;
	int err, tmpval;

	valptr = (uint16_t *)arg1 + cpu;
	tmpval = *valptr;

	err = sysctl_handle_int(oidp, &tmpval, 0, req);
	if (err || req->newptr == NULL)
		return (err);

	if ((tmpval<0) || (tmpval>1024))
		return (EINVAL);

	*valptr = (uint16_t)(tmpval & 0xffff);

	hp = DPCPU_ID_PTR(cpu, hmp_pcpu);
	if ((uint16_t *)arg1 == cidx_inst.score.performance)
		hmp_set_score(hp, HMP_SCORE_PERF, tmpval);
	else if ((uint16_t *)arg1 == cidx_inst.score.efficiency)
		hmp_set_score(hp, HMP_SCORE_EFF, tmpval);

	return (0);
}

static int
coreindex_score_register_sysctl(void)
{
	char desc[128];

	/* Set sysctl capacity value */
	for (int i=0; i<mp_ncpus; i++) {
		snprintf(desc, sizeof(desc), "CPU core #%d performance score index", i);
		SYSCTL_ADD_PROC(&(cidx_inst.sysctl.ctx), SYSCTL_CHILDREN(*(cidx_inst.sysctl.core+i)),
				OID_AUTO, "performance", CTLTYPE_INT | CTLFLAG_RW | CTLFLAG_MPSAFE,
				(void *)(cidx_inst.score.performance), i, coreindex_score_set_sysctl,"I", desc);

		snprintf(desc, sizeof(desc), "CPU core #%d efficiency score index", i);
		SYSCTL_ADD_PROC(&(cidx_inst.sysctl.ctx), SYSCTL_CHILDREN(*(cidx_inst.sysctl.core+i)),
				OID_AUTO, "efficiency", CTLTYPE_INT | CTLFLAG_RW | CTLFLAG_MPSAFE,
				(void *)(cidx_inst.score.efficiency), i, coreindex_score_set_sysctl,"I", desc);
	}

	return (0);
}

